File: Go/Day15.go
package main

import AH "./adventhelper"

import (

)

func MakeGameState(inputs []int) (gameState map[int]int) {
	gameState = make(map[int]int)
	for i, v :=	range inputs {
		gameState[v] = i + 1
	}
	return
}

func playGame(game map[int]int, n int, seen bool, turns int, offset int) (next int) {
	for turn := offset; turn <= turns; turn++ {
		if seen {
 			next = turn - game[n] - 1
		} else { // if not say zero
			next = 0
		}
		_, seen = game[next]
		game[n] = turn - 1
		n = next
	}

	return
}

func main() {
	input15 := []int{7,14,0,17,11,1,2}

	q := MakeGameState(input15)
	init := 0
	if _, ok := q[0]; ok {
		init = input15[len(input15) - 1]
	}
	s := playGame(q, init, false, 2020, len(input15) + 1)

	q = MakeGameState(input15) // need to reset the map
	t := playGame(q, init, false, 30000000, len(input15) + 1)

	AH.PrintSoln(15, s, t)

	return
}

File: C++/Day15.cpp
#include "AH.h"

namespace Day15
{
	std::map<uint64_t, uint64_t> makeGameState(const std::vector<uint64_t>& inputs)
	{
		std::map<uint64_t, uint64_t> gameState;
		for (uint64_t i = 0; i < inputs.size(); ++i)
			gameState[inputs[i]] = i + 1;

		return gameState;
	}

	uint64_t playGame(std::map<uint64_t, uint64_t>& game, uint64_t n,
	                  bool seen, const uint64_t turns, const uint64_t offset)
	{
		uint64_t next = 0;
		for (uint64_t turn = offset; turn <= turns; ++turn)
		{
			if (seen)
				next = turn - game[n] - 1;
			else
				next = 0;

			seen = ( game.find(next) != game.end() );
			game[n] = turn - 1;
			n = next;
		}

		return n;
	}

	int Run(const std::string& filename)
	{
		std::vector<uint64_t> input{7,14,0,17,11,1,2};
		auto q1 = makeGameState(input);
		auto q2 = makeGameState(input);
		uint64_t init = 0;

		bool seenZero = ( q1.find(0) != q1.end() );
		if (seenZero)
			init = input.back();

		auto s = playGame(q1, init, seenZero, 2020,     input.size() + 1);
		auto t = playGame(q2, init, seenZero, 30000000, input.size() + 1);

		AH::PrintSoln(15, s, t);

		return 0;
	}

}
